### DISCUSSION TITLE
catch case as a shorthand for catch and switch case

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#2688] [Language Ideas]
As of now C# `switch` syntax has extended to support many kind of expression. I think it was so overlapped with the old `catch` syntax

```C#
try
{
}
catch(ExceptionA ea)
{
}
catch(ExceptionB eb)
{
}
catch(ExceptionC ec)
{
}
catch(Exception e)
{
    switch(e)
    {
        default:
            throw e;
    }
}
```

It actually could replace with switch

```C#
try
{
}
catch(Exception e)
{
    switch(e)
    {
        case ExceptionA ea:
            break;
        case ExceptionB eb:
            break;
        case ExceptionC ec:
            break;
        default:
            throw e;
    }
}
```

And I think we could also make the shorthand syntax for convenient

```C#
try
{
}
catch(e) // Exception e // dropping extension type as a syntax for switch mode
{
    case ExceptionA ea:
        break;
    case ExceptionB eb:
        break;
    case ExceptionC ec:
        break;
    default:
        throw e;
}
```

The advantage of `switch` is it could fall through and could specify additional `when`. And it will support all kind of `switch` expression today and future, we don't need to extend support to `catch` syntax. It just transpiled into `switch`

**Comment by yaakov-h** [comment]
> The advantage of `switch` is it could ... specify additional `when`

How is this an advantage over exception filters (`catch (Exception ex) when (...)`)?

**Comment by Thaina** [comment]
@yaakov-h Sorry I use the wrong words there. What I meant actually is `case pattern`

But also I wonder that `catch-when` would always be the same with `case-when` forever. Will there be any difference? Or we would always make `catch` and `case` always be the same?

**Comment by yaakov-h** [comment]
`catch-when` is a fault, as is executed when the exception is thrown.
`case-when` as it is today would only be executed when the exception is caught, and any syntactic sugar would likely keep that behaviour.
